File: src/Enterprise/Kotlin/Example2.kt
// Abstract Product
interface Database {
    fun connect()
    fun query()
}

// Concrete Product
class Postgres : Database {
    override fun connect() {
        println("Connecting to PostgreSQL")
    }

    override fun query() {
        println("Querying PostgreSQL")
    }
}

class MySQL : Database {
    override fun connect() {
        println("Connecting to MySQL")
    }

    override fun query() {
        println("Querying MySQL")
    }
}

// Abstract Factory
interface DatabaseFactory {
    fun createDatabase(): Database
}

// Concrete Factory
class PostgresFactory : DatabaseFactory {
    override fun createDatabase(): Database {
        return Postgres()
    }
}

class MySQLFactory : DatabaseFactory {
    override fun createDatabase(): Database {
        return MySQL()
    }
}

fun useDatabase(factory: DatabaseFactory) {
    val db = factory.createDatabase()
    db.connect()
    db.query()
}

// Usage
fun main() {
    useDatabase(PostgresFactory())
    useDatabase(MySQLFactory())
}

File: src/Other/MicroPython/example2.py
class Database:
    def connect(self):
        pass

    def query(self):
        pass

class Postgres(Database):
    def connect(self):
        print("Connecting to PostgreSQL")

    def query(self):
        print("Querying PostgreSQL")

class MySQL(Database):
    def connect(self):
        print("Connecting to MySQL")

    def query(self):
        print("Querying MySQL")

class DatabaseFactory:
    def create_database(self):
        pass

class PostgresFactory(DatabaseFactory):
    def create_database(self):
        return Postgres()

class MySQLFactory(DatabaseFactory):
    def create_database(self):
        return MySQL()

def use_database(factory):
    db = factory.create_database()
    db.connect()
    db.query()

# Usage
use_database(PostgresFactory())
use_database(MySQLFactory())
